### PR TITLE
feat(rotation): add zero-downtime key rotation utility

### PR DESCRIPTION
**Pull request by Alberto-Codes**
Before this PR, same-backend passphrase rotation (e.g., rotating a Fernet key) had no supported path — `additional_backends` raises `ConfigurationError` for duplicate `backend_id` values. This adds `rotate_encryption_keys()`, a standalone async utility that re-encrypts all session data directly via SQLAlchemy, bypassing the TypeDecorator to control which key decrypts.

- Add `rotate_encryption_keys(db_url, old_backend, new_backend) -> RotationResult` in new `rotation.py` module; reads raw base64 TEXT, parses envelope, decrypts with `old_backend`, re-encrypts with `new_backend`, writes back with `UPDATE WHERE update_time = :read_time` optimistic concurrency guard
- Add `RotationResult(rotated, skipped)` dataclass; re-runs are idempotent (already-rotated records carry new `backend_id` and are skipped silently)
- Export both from public API (`__init__.py` `__all__`)
- Add ADR-009 documenting two rotation paths (batch vs lazy), no-coordinator decision, `update_time` vs `version` column rationale
- Add `docs/how-to/key-rotation.md` operator guide with Path A (lazy cross-backend via `additional_backends`) and Path B (batch via `rotate_encryption_keys()`) examples
- Add 10 unit tests (mocked engine, covers NULL column, optimistic concurrency, backend-id filtering, error safety) and 6 integration tests (real SQLite + real FernetBackend keys)

Test: `uv run pytest tests/unit/test_rotation.py tests/integration/test_key_rotation.py -v`

---

### Checklist
- [x] Self-reviewed my code
- [x] Tests pass (`uv run pytest`)
- [x] Lint passes (`uv run ruff check .`)
- [ ] Breaking changes use `!` in title and `BREAKING CHANGE:` in body

### Review Focus
- `rotation.py:_rotate_table` — raw SQL SELECT/UPDATE loop; per-record `asyncio.to_thread()` crypto; optimistic concurrency via `update_time`; events table `rowcount=0` treated as cascade-deleted (not skipped)
- `rotation.py:rotate_encryption_keys` — four separate transactions (one per table); re-run idempotency contract in docstring
- Party pre-review consensus: NULL `event_data` test, docstring idempotency/events semantics, and scalability note all addressed before this PR

### Related
- Story 3.3 (multi-backend coexistence) — Path A foundation (`additional_backends`)
- ADR-007 (architecture migration) — why `update_time` is available as concurrency guard

### CODE REVIEW

File: src/adk_secure_sessions/rotation.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    ``new_backend``. Records already on a different backend are skipped
+    silently.
+
+    Uses ``update_time`` as an optimistic concurrency guard for
+    ``sessions``, ``app_states``, and ``user_states``. If a record is
+    modified between the rotation function's read and write
+    (``rows_affected == 0``), it is counted as skipped. Run the function
+    again to pick up remaining records. Re-runs are safe: records already
+    rotated carry ``new_backend.backend_id`` in their envelope and are
+    skipped silently, so calling this function multiple times is idempotent.
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply src/adk_secure_sessions/rotation.py]: Addressed in latest push — scoped the idempotency claim: re-runs are safe for cross-backend rotation (backend IDs differ, already-rotated records are skipped by the envelope filter) but not for same-backend rotation (same backend_id, old key cannot decrypt new-key ciphertext). Docstring and how-to guide updated to make this distinction explicit.

File: src/adk_secure_sessions/rotation.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+                _sync_reencrypt,
+                ciphertext,
+                old_backend.sync_decrypt,
+                new_backend.sync_encrypt,
+            )
+        except DecryptionError:
+            raise
+        except Exception as exc:
+            msg = f"Rotation failed: re-encryption error in table {table!r}"
+            raise DecryptionError(msg) from exc
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply src/adk_secure_sessions/rotation.py]: Addressed — changed to `from None`. Consistent with the existing codebase pattern: all `raise DecryptionError` sites in `backends/` and `type_decorator.py` already use `from None` for the same reason. Also removed the now-unused `as exc` binding (ruff F841).

File: src/adk_secure_sessions/rotation.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+
+    # Build SELECT — column names are hardcoded constants, not user input
+    select_col_list = pk_cols + [enc_col]
+    if has_update_time:
+        select_col_list = select_col_list + ["update_time"]
+    select_cols = ", ".join(select_col_list)
+    select_sql = f"SELECT {select_cols} FROM {table}"
+    rows = (await conn.execute(text(select_sql))).fetchall()
+
+    for row in rows:
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply src/adk_secure_sessions/rotation.py]: Keeping as-is — `fetchall()` is intentional for this maintenance utility scope. The scalability limitation is documented in ADR-009 and the function docstring ("run during a low-traffic window"). Streaming iteration via `conn.stream()` is a planned enhancement for operators with very large databases and would be addressed in a follow-up story.

File: docs/how-to/key-rotation.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  exception messages from `rotate_encryption_keys()`.
+- The old key is not retained anywhere after the function returns — it is the
+  caller's responsibility to securely delete or revoke the old key/passphrase.
+- During rotation, the database contains a mix of old-key and new-key records.
+  If the service is running, it must continue to use `old_backend` for reads
+  until rotation completes and the service is reconfigured.
+
+---
+
+## See Also
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply docs/how-to/key-rotation.md]: Addressed — renamed `## See Also` to `## Related` to match `docs/faq.md` and `docs/getting-started.md` convention.

File: tests/integration/test_key_rotation.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        assert result1.rotated > 0
+
+        # Second pass with same old/new: for same backend_id, records are
+        # visible again (both Fernet = 0x01). They will be decrypted with
+        # new_backend's key (fails) ... so second pass should fail gracefully
+        # OR: use new_backend as old_backend to rotate back — confirm it works
+        result2 = await rotate_encryption_keys(
+            db_url=db_url, old_backend=new_backend, new_backend=old_backend
+        )
+        # Records were re-encrypted to new key; rotating back to old key
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply tests/integration/test_key_rotation.py]: Addressed — renamed to `test_can_rotate_back_between_fernet_backends` with a rewritten docstring accurately describing the reverse-rotation scenario. Also clarified in the docstring why re-running with the original old_backend is not the correct pattern for same-backend rotation recovery.

File: src/adk_secure_sessions/rotation.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+      new_backend.backend_id``): Re-runs are safe. Already-rotated records
+      carry ``new_backend.backend_id`` in their envelope and are skipped
+      silently by the backend-id filter.
+    - *Same-backend rotation* (``old_backend.backend_id ==
+      new_backend.backend_id``, e.g., two ``FernetBackend`` instances):
+      A single pass is expected. Re-running with the original ``old_backend``
+      will attempt to decrypt already-rotated ciphertext with the old key
+      and raise ``DecryptionError``. For same-backend rotation, run once,
+      then update your service configuration to use ``new_backend``.
+
```
[the remark this answers is not in the document]
Comment (Alberto-Codes) [reply src/adk_secure_sessions/rotation.py]: Addressed — added explicit pre-condition to the rotation.py docstring: stop or pause the service before running same-backend rotation to prevent read failures from the mixed-key database state.
